src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+(define-constant ERR_NAME_GRACE_PERIOD 2009)
+(define-constant ERR_NAME_BLANK 2010)
+(define-constant ERR_NAME_ALREADY_CLAIMED 2011)
+(define-constant ERR_NAME_CLAIMABILITY_EXPIRED 2012)
+(define-constant ERR_NAME_NOT_FOUND 2013)
+(define-constant ERR_NAME_REVOKED 2014)
+(define-constant ERR_NAME_TRANSFER_FAILED 2015)
+(define-constant ERR_NAME_PREORDER_ALREADY_EXISTS 2016)
+(define-constant ERR_NAME_HASH_MALFORMED 2017)
+(define-constant ERR_NAME_PREORDERED_BEFORE_NAMESPACE_LAUNCH 2018)
+(define-constant ERR_NAME_NOT_RESOLVABLE 2019)
+(define-constant ERR_NAME_COULD_NOT_BE_MINTED 2020)
+(define-constant ERR_NAME_COULD_NOT_BE_TRANSFERED 2021)
+(define-constant ERR_NAME_CHARSET_INVALID 2022)
+
+(define-constant ERR_PRINCIPAL_ALREADY_ASSOCIATED 3001)
+(define-constant ERR_INSUFFICIENT_FUNDS 4001)
+
+;;;; Constants
+(define-constant BURN_ADDRESS 'S0000000000000000000002AA028H)
+
+;; TTL
+;; todo(ludo): add real-life values
+(define-constant NAMESPACE_PREORDER_CLAIMABILITY_TTL u10)
+(define-constant NAMESPACE_LAUNCHABILITY_TTL u10)
+(define-constant NAME_PREORDER_CLAIMABILITY_TTL u10)
+(define-constant NAME_GRACE_PERIOD_DURATION u5)
+
+(define-constant ATTACHMENTS_INV_PAGE_SIZE u8)
+(define-data-var attachments-inv-index-cursor uint u0)
+(define-data-var attachments-inv-page-cursor uint u0)
+(define-map attachments-inv 
+    ((page uint) (index uint)) 
+    ((content-hash (buff 20))))
+
+;; Price tables
+(define-constant NAMESPACE_PRICE_TIERS (list
+  u96000 
+  u9600 u9600 
+  u960 u960 u960 u960 
+  u96 u96 u96 u96 u96 u96 u96 u96 u96 u96 u96 u96 u96))
+
+;;;; Data
+(define-map namespaces
+  ((namespace (buff 19)))
+  ((namespace-import principal)
+   (revealed-at uint)
+   (launched-at (optional uint))
+   (namespace-version uint)
+   (renewal-rule uint)
+   (price-function (tuple 
+    (buckets (list 16 uint)) 
+    (base uint) 
+    (coeff uint) 
+    (nonalpha-discount uint) 
+    (no-vowel-discount uint)))))
+
+(define-map namespace-preorders
+  ((hashed-salted-namespace (buff 20)) (buyer principal))
+  ((created-at uint) (claimed bool) (stx-burned uint)))
+
+(define-non-fungible-token names (tuple (name (buff 16)) (namespace (buff 19))))
+
+(define-map owner-name ((owner principal)) ((name (buff 16)) (namespace (buff 19))))
+
+(define-map name-properties
+  ((name (buff 16)) (namespace (buff 19)))
+  ((registered-at (optional uint))
+   (imported-at (optional uint))
+   (revoked-at (optional uint))
+   (zonefile-hash (buff 20))))
+
+(define-map name-preorders
+  ((hashed-salted-fqn (buff 20)) (buyer principal))
+  ((created-at uint) (claimed bool) (stx-burned uint)))
+
+(define-private (min (a uint) (b uint))
+  (if (<= a b) a b))
+
+(define-private (max (a uint) (b uint))
+  (if (> a b) a b))
+
+(define-read-only (compute-namespace-price? (namespace (buff 19)))
+  (let ((namespace-len (len namespace)))
+    (asserts!
+      (> namespace-len u0)
+      (err ERR_NAMESPACE_BLANK))
+    (ok (get value (fold 
```

Comment:
Is there an issue somewhere in Clarity to look up the _ith_ item in a list?

Reply:
I don't think so. We talked about adding this a while ago, and think we settled on a **no**.

Reply:
Ah, I see.  Is there a conversation online somewhere on this?  I can't remember why we decided this (cc @kantai).

Reply:
If I recall correctly, our reasoning was that if it was something that was implementable as a "library" function, then we shouldn't implement it as a native function. However, `element-at` isn't really implementable as a library function, because the output type depends on the input list type, so I think our conclusion from then is probably wrong.

Reply:
Went ahead an added an issue for this here: https://github.com/clarity-lang/reference/issues/39.  Let me know if it should be added in this repo as well?